README.pod

=pod

=head1 NAME

MojoX::JSON::RPC - a Perl implementation of the JSON-RPC 2.0 protocol for Mojolicious

=head1 SYNOPSIS

Server as plugin (L<Mojolicious::Plugin::JsonRpcDispatcher>):

    #!/usr/bin/env perl
    use Mojolicious::Lite;
    use MojoX::JSON::RPC::Service;

    plugin 'json_rpc_dispatcher' => {
        services => {
            '/jsonrpc' => MojoX::JSON::RPC::Service->new->register(
                'sum',
                sub {
                    my @params = @_;
                    my $sum    = 0;
                    $sum += $_ for @params;
                    return $sum;
                }
            )
        }
    };

    app->start;

Client (L<MojoX::JSON::RPC::Client>):

    #!/usr/bin/env perl
    use MojoX::JSON::RPC::Client;

    my $client = MojoX::JSON::RPC::Client->new;
    my $url    = 'http://www.example.com/jsonrpc';
    my $callobj = {
        id      => 1,
        method  => 'sum',
        params  => [ 17, 25 ]
    };

    my $res = $client->call($url, $callobj);

    if($res) {
        if ($res->is_error) { # RPC ERROR
            print 'Error : ', $res->error_message;
        }
        else {
            print $res->result;
        }
    }
    else {
        my $tx_res = $client->tx->res; # Mojo::Message::Response object
        print 'HTTP response '.$tx_res->code.' '.$tx_res->message;
    }

Non-blocking client:

    #!/usr/bin/env perl
    use MojoX::JSON::RPC::Client;

    my $client = MojoX::JSON::RPC::Client->new;
    my $url    = 'http://www.example.com/jsonrpc';
    my $callobj = {
        id      => 1,
        method  => 'sum',
        params  => [ 17, 25 ]
    };

    $client->call($url, $callobj, sub {
        # With callback
        my $res = pop;

        if($res) {
            if ($res->is_error) { # RPC ERROR
                print 'Error : ', $res->error_message;
            }
            else {
                print $res->result;
            }
        }
        else {
           my $tx_res = $client->tx->res; # Mojo::Message::Response object
           print 'HTTP response '.$tx_res->code.' '.$tx_res->message;
        }

        Mojo::IOLoop->stop;
    });

    Mojo::IOLoop->start;

=head1 DESCRIPTION

This module implments a client plugin and a server plugin for JSON-RPC 2.0 for
use with Mojolicious.

This module follows the draft specficiation for JSON-RPC 2.0. More information
can be found at L<http://groups.google.com/group/json-rpc/web/json-rpc-2-0>.

=head1 SEE ALSO

L<Mojolicious::Plugin::JsonRpcDispatcher>, L<MojoX::JSON::RPC::Dispatcher>,
L<MojoX::JSON::RPC::Client>

=head1 AUTHOR

Henry Tang

=head1 CREDITS

=over 2

Igor Afanasyev

Renee Baecker (https://github.com/reneeb)

=back

=head1 COPYRIGHT & LICENSE

Copyright (C) 2011-2015, Henry Tang.

MojoX::JSON::RPC is provided "as is" and without any express or
implied warranties, including, without limitation, the implied warranties
of merchantibility and fitness for a particular purpose.

This program is free software, you can redistribute it and/or modify it
under the terms of the Artistic License version 2.0.

=cut


Changes
This file documents the revision history for Perl extension MojoX::JSON::RPC.

0.09 2016-05-24 00:00:00
     - is_debug is deprecated and removed from Mojo::Log (Vidar Tyldum <vidar@tyldum.com>)
     - Updated copyright to 2016

0.08 2015-02-08 00:00:00
     - Fixed typo in credit
     - Bumped version. 0.0.7 was packaged incorrectly.

0.07 2015-02-07 00:00:00
     - Replace OO-interface of Mojo::JSON with functional interface (reneeb)
     - Add metaresources to Makefile.PL (reneeb)
     - Updated PREREQ_PM to Mojolicious 5.77
     - Updated copyright to 2015

0.06 2014-02-02 00:00:00
     - Fixed: #87793: The client cannot be used independently
     - Fixed: #92613: Deprecated use of Mojo::UserAgent::app
     - Fixed: Deprecated use of Mojolicious::secret
     - Updated PREREQ_PM to Mojolicious 4.74
     - Updated copyright to 2014

0.05 2013-05-26 00:00:00
     - remove use of render_json to make Mojolicious 4.0 happy
     - Updated PREREQ_PM to Mojolicious 4.00

0.04 2012-05-14 00:00:00
     - Added 'with_self' option which gives access to the controller object itself (iafan)
     - Added tests for 'with_self'

0.03 2012-02-20 00:00:00
     - Fixed: log attribute is removed from Mojo::UserAgent in Mojolicious 2.39.

0.02 2011-12-22 00:00:00
     - Updated req. to Mojolicious 1.86 and fixed some deprecation warnings in tests

0.01 2011-05-07 00:00:00
     - First release.


lib/MojoX/JSON/RPC/Client.pm
package MojoX::JSON::RPC::Client;

use Mojo::Base -base;
use Mojo::JSON qw(encode_json decode_json);
use Mojo::UserAgent;

has id           => undef;
has ua           => sub { Mojo::UserAgent->new };
has version      => '2.0';
has content_type => 'application/json';
has tx           => undef;                          # latest transaction

sub call {
    my ( $self, $uri, $body, $callback ) = @_;

    # body might be json already, only encode if necessary
    if ( ref $body eq 'HASH' || ref $body eq 'ARRAY' ) {
        foreach my $o ( ref $body eq 'HASH' ? $body : @{$body} ) {
            $o->{version} ||= $self->version;
        }
        $body = encode_json($body);
    }
    else {
        $body ||= q{};
    }

    # Always POST if $body is not empty!
    if ( ref $callback ne 'CODE' ) {
        if ( $body ne q{} ) {
            return $self->_process_result(
                $self->ua->post(
                    $uri, { Content_Type => $self->content_type }, $body
                )
            );
        }
        elsif ( $uri =~ /\?/xms ) {
            return $self->_process_result( $self->ua->get($uri) );
        }
    }
    else {    # non-blocking
        if ( $body ne q{} ) {
            $self->ua->post(
                $uri,
                { Content_Type => $self->content_type },
                $body,
                sub {    # callback
                    $callback->( $self->_process_result(pop) );
                },
            );
            return;
        }
        elsif ( $uri =~ /\?/xms ) {
            $self->ua->get(
                $uri => sub {    # callback
                    $callback->( $self->_process_result(pop) );
                }
            );
            return;
        }
    }
    return Carp::croak 'Cannot process call!';
}

# Prepare a Proxy object
sub prepare {
    my $self = shift;

    my %m = ();
URI:
    while ( my $uri = shift ) {
        my $methods = shift;

        # methods can be a name, a reference to a name or
        # a reference to an array of names
        if ( ref $methods eq 'SCALAR' ) {
            $methods = [$$methods];
        }
        elsif ( defined $methods && ref $methods eq q{} ) {
            $methods = [$methods];
        }
        if ( ref $methods ne 'ARRAY' ) {
            last URI;
        }
    METHOD:
        foreach my $method ( @{$methods} ) {
            if ( exists $m{$method} && $m{$method} ne $uri ) {
                Carp::croak qq{Cannot register method $method twice!};
            }
            $m{$method} = $uri;
        }
    }
    return bless {
        client  => $self,
        methods => \%m
        },
        'MojoX::JSON::RPC::Client::Proxy';
}

sub _process_result {
    my ( $self, $tx ) = @_;

    $self->tx($tx);    # save latest transaction

    my $tx_res = $tx->res;
    my $log = $self->ua->server->app->log if $self->ua->server->app;
    if ( $log && $log->is_debug ) {
        $log->debug( 'TX BODY: [' . $tx_res->body . ']' );
    }

    # Check if RPC call is succesfull
    if ( !( $tx_res->is_status_class(200) ) )
    {
        return;
    }

    my $decode_error;
    my $rpc_res;
    
    eval{ $rpc_res = decode_json( $tx_res->body || '{}' ); 1; } or $decode_error = $@;
    if ( $decode_error && $log ) {    # Server result cannot be parsed!
        $log->error( 'Cannot parse rpc result: ' . $decode_error );
        return;
    }

    # Return one or more ReturnObject's
    return ref $rpc_res eq 'ARRAY'
        ? [
        map {
            MojoX::JSON::RPC::Client::ReturnObject->new( rpc_response => $_ )
        } ( @{$rpc_res} )
        ]
        : MojoX::JSON::RPC::Client::ReturnObject->new(
        rpc_response => $rpc_res );
}

package MojoX::JSON::RPC::Client::Proxy;

use Carp;
use warnings;
use strict;

# no constructor defined. Object creation
# done by MojoX::JSON::RPC::Client.

our $AUTOLOAD;

# Dispatch calls
sub AUTOLOAD {
    my $self = shift;

    my $method = $AUTOLOAD;
    $method =~ s/.*:://;

    # We do not want to overload DESTROY
    if ( $method eq 'DESTROY' ) {
        return;
    }

    if ( !exists $self->{methods}->{$method} ) {
        Carp::croak "Unsupported method $method";
    }

    my $res = $self->{client}->call(
        $self->{methods}->{$method},
        {   id     => $self->{id}++,
            method => $method,
            params => \@_
        }
    );
    return defined $res ? $res->result : ();
}

package MojoX::JSON::RPC::Client::ReturnObject;

use Mojo::Base -base;

has rpc_response => undef;    # rpc response

sub result {
    my ($self) = @_;
    my $rpc_response = $self->rpc_response;
    return
        ref $rpc_response eq 'HASH' && exists $rpc_response->{result}
        ? $rpc_response->{result}
        : undef;
}

sub id {
    my ($self) = @_;
    my $rpc_response = $self->rpc_response;
    return
        ref $rpc_response eq 'HASH' && exists $rpc_response->{id}
        ? $rpc_response->{id}
        : undef;
}

sub is_error {
    my ($self) = @_;
    my $rpc_response = $self->rpc_response;
    return ref $rpc_response eq 'HASH' && exists $rpc_response->{error}
        ? 1
        : 0;
}

sub error_code {
    my ($self) = @_;
    return $self->is_error ? $self->rpc_response->{error}->{code} : undef;
}

sub error_message {
    my ($self) = @_;
    return $self->is_error ? $self->rpc_response->{error}->{message} : undef;
}

sub error_data {
    my ($self) = @_;
    return $self->is_error ? $self->rpc_response->{error}->{data} : undef;
}

1;

__END__

=head1 NAME

MojoX::JSON::RPC::Client - JSON RPC client

=head1 SYNOPSIS

    use MojoX::JSON::RPC::Client;

    my $client = MojoX::JSON::RPC::Client->new;
    my $url    = 'http://www.example.com/jsonrpc/API';
    my $callobj = {
        id      => 1,
        method  => 'sum',
        params  => [ 17, 25 ]
    };

    my $res = $client->call($url, $callobj);

    if($res) {
        if ($res->is_error) { # RPC ERROR
            print 'Error : ', $res->error_message;
        }
        else {
            print $res->result;
        }
    }
    else {
        my $tx_res = $client->tx->res; # Mojo::Message::Response object
        print 'HTTP response '.$tx_res->code.' '.$tx_res->message;
    }

Non-blocking:

    $client->call($url, $callobj, sub {
        # With callback
        my $res = pop;

        # ... process result ...

        Mojo::IOLoop->stop;
    });

    Mojo::IOLoop->start;

Easy access:

    my $proxy = $client->prepare($uri, ['sum', 'echo']);

    print $proxy->sum(10, 23);


=head1 DESCRIPTION

A JSON-RPC client.

=head1 ATTRIBUTES

L<MojoX::JSON::RPC::Client> implements the following attributes.

=head2 C<id>

Id used for JSON-RPC requests. Used when no id is provided as request parameter.

=head2 C<ua>

L<Mojo::UserAgent> object.

=head2 C<json>

L<Mojo::JSON> object for encoding and decoding.

=head2 C<version>

JSON-RPC version. Defaults to 2.0.

=head2 C<content_type>

Content type. Defaults to application/json.

=head2 C<tx>

Mojo::Transaction object of last request.

=head1 METHODS

L<MojoX::JSON::RPC::Client> inherits all methods from L<Mojo::Base> and implements the
following new ones.

=head2 C<new>

Creates new MojoX::JSON::RPC::Client object.

    my $client = MojoX::JSON::RPC::Client->new;

=head2 C<call>

Execute JSON-RPC call. Returns L<MojoX::JSON::RPC::CLient::ReturnObject> if RPC call
is executed correctly.

    my $client = MojoX::JSON::RPC::Client->new;
    my $url    = 'http://www.example.com/jsonrpc/API';
    my $callobj = {
        id      => 1,
        method  => 'sum',
        params  => [ 17, 25 ]
    };

    my $res = $client->call($url, $callobj);
    if($res) {
        if ($res->is_error) { # RPC error
            print 'Error : ', $res->error_message;
        }
        else {
            print $res->result;
        }
    }
    else {
        my $tx_res = $client->tx->res; # Mojo::Message::Response object
        print 'HTTP response '.$tx_res->code.' '.$tx_res->message;
    }

Make non-blocking call:

    $client->call($url, $callobj, sub {
        # With callback
        my $res = pop;

        # ... process result ...

        Mojo::IOLoop->stop;
    });

    Mojo::IOLoop->start;

=head2 C<prepare>

Prepares a proxy object that allows RPC methods to be called
more easily.

    my $proxy = $client->prepare($uri, ['sum', 'echo']);

    my $res = $proxy->sum(1, 2);

    print $proxy->echo("Echo this!");

Register services from multiple urls at once:

    my $proxy = $client->prepare($uri1, 'sum', $uri2, [ 'echo', 'ping' ]);

    my $res = $proxy->sum(1, 2);

    print $proxy->echo("Echo this!");

    my $ping_res = $proxy->ping;

=head1 C<MojoX::JSON::RPC::CLient::ReturnObject>

This object is returned by C<call>.

=head2 C<result>

RPC result.

=head2 C<is_error>

Returns a boolean indicating whether an error code has been set.

=head2 C<error_code>

RPC error code.

=head2 C<error_message>

RPC error message.

=head2 C<error_data>

RPC error data.

=head1 SEE ALSO

L<MojoX::JSON::RPC>

=cut



lib/MojoX/JSON/RPC/Service.pm
package MojoX::JSON::RPC::Service;

use Mojo::Base -base;

{

    # Store rpc methods registered by register_rpc_method_names
    my $_rpcs = undef;

    sub new {
        my $class = shift;
        my $self  = $class->SUPER::new(@_);

        # make shallow copy of class level rpcs
        if ( ref $_rpcs eq 'HASH' && exists $_rpcs->{$class} ) {
            $self->{_rpcs} = { %{ $_rpcs->{$class} } };
        }
        return $self;
    }

    sub register_rpc_method_names {
        my ( $class, @methods ) = @_;

    METHOD:
        foreach my $m (@methods) {
            my $name   = $class . '::' . $m;
            my $method = eval qq|*{*$name}{CODE}|;

            if ( ref $method ne 'CODE' ) {
                Carp::croak
                    qq{register_rpc_method_names: $name not a method.};
            }
            if ( exists $_rpcs->{$class}->{$m} ) {
                Carp::croak
                    qq{register_rpc_method_names: $name already registered.};
            }
            $_rpcs->{$class}->{$m} = { method => $method, with_svc_obj => 1 };
        }

        return $class;
    }
}

sub register {
    my ( $self, $name, $sub, $options ) = @_;

    if ( !defined $name || $name eq q{} ) {
        Carp::croak 'name?';
    }
    if ( ref $sub ne 'CODE' ) {
        Carp::croak qq{name[$name] code?};
    }
    if ( exists $self->{_rpcs}->{$name} ) {
        Carp::croak qq{$name already registered};
    }

    $options ||= {};
    if ( ref $options ne 'HASH' ) {
        Carp::croak 'options?';
    }

    my %obj = ( method => $sub );
OPTION:
    foreach my $opt ( 'with_mojo_tx', 'with_svc_obj' ) {
        if ( !exists $options->{$opt} ) {
            next OPTION;
        }
        $obj{$opt} = $options->{$opt};
    }

    $self->{_rpcs}->{$name} = \%obj;
    return $self;
}

1;

__END__

=head1 NAME

MojoX::JSON::RPC::Service - JSON RPC Service registration

=head1 SYNOPSIS

    use MojoX::JSON::RPC::Service;

    my $svc  = MojoX::JSON::RPC::Service->new;

    $svc->register(
        'sum',
        sub {
            my @params = @_;
            my $sum    = 0;
            $sum += $_ for @params;
            return $sum;
        }
    );

    $svc->register(
        'remote_address',
        sub {
            my $tx = shift;
            return $tx->remote_address;
        },
        {
             with_mojo_tx => 1
        }
    );


    ## Then in Mojolicious application
    $self->plugin(
        'json_rpc_dispatcher',
        services => {
            '/jsonrpc'  => $svc,
        }
    );

This package can also be used as a base class to make it easy to create object-oriented
JSON-RPC applications:

    package MyService;

    use Mojo::Base 'MojoX::JSON::RPC::Service';

    sub sum {
        my ($self, @params) = @_;
        my $sum    = 0;
        $sum += $_ for @params;
        return $sum;
    }

    __PACKAGE__->register_rpc_method_names( 'sum' );

    ## Then in Mojolicious application
    $self->plugin(
        'json_rpc_dispatcher',
        services => {
            '/jsonrpc'  => MyService->new,
        }
    );

=head1 DESCRIPTION

Register JSON-RPC service calls.

=head1 METHODS

L<MojoX::JSON::RPC::Service> inherits all methods from L<Mojo::Base> and implements the
following new ones.

=head2 C<register>

Register RPC methods.

    $svc->register(
        'sum',
        sub {
            my @params = @_;
            my $sum    = 0;
            $sum += $_ for @params;
            return $sum;
       }
    );

with_mojo_tx can be passed as options. In that case, L<Mojo::Transaction>
object will be pass as first argument of the subroutine.

    $svc->register(
        'remote_address',
        sub {
            my $tx = shift;
            return $tx->remote_address;
        },
        {
             with_mojo_tx => 1
        }
    );

=head2 C<register_rpc_method_names>

Class method. Register a list of methods as JSON-RPC calls.

    __PACKAGE__->register_rpc_method_names( 'sum', 'multiply' );

=head1 SEE ALSO

L<MojoX::JSON::RPC::Dispatcher>

=cut


lib/MojoX/JSON/RPC/Dispatcher.pm
package MojoX::JSON::RPC::Dispatcher;

use Mojo::Base 'Mojolicious::Controller';
use Mojo::JSON qw(decode_json);
use MojoX::JSON::RPC::Dispatcher::Method;
use MojoX::JSON::RPC::Service;

# process JSON-RPC call
sub call {
    my ($self) = @_;

    my $rpc_response = $self->_handle_methods( $self->_acquire_methods );
    if ( !$rpc_response ) {    # is notification
        $self->tx->res->code(204);
        return $self->rendered;
    }
    my $res = $self->tx->res;
    $res->code(
        $self->_translate_error_code_to_status(
            ref $rpc_response eq 'HASH' && exists $rpc_response->{error}
            ? $rpc_response->{error}->{code}
            : q{}
        )
    );
    $res->headers->content_type('application/json-rpc');
    return $self->render(json => $rpc_response);
}

sub _acquire_methods {
    my ($self) = @_;

    my $log    = $self->app->log;
    my $req    = $self->req;        # Mojo::Message::Request object
    my $method = $req->method;      # GET / POST
    my $request;

    if ( $method eq 'POST' ) {
        if ( $log->is_debug ) {
            $log->debug( 'REQUEST: BODY> ' . $req->body );
        }
        
        my $decode_error;
        eval{ $request = decode_json( $req->body ); 1; } or $decode_error = $@;
        if ( $decode_error ) {
            $log->debug( 'REQUEST: JSON error> ' . $decode_error );
            return MojoX::JSON::RPC::Dispatcher::Method->new->parse_error(
                $decode_error );
        }
    }
    elsif ( $method eq 'GET' ) {
        my $params = $req->query_params->to_hash;
        my $decoded_params;

        if ( exists $params->{params} ) {
            my $decode_error;
            eval{ $decoded_params = decode_json( $params->{params} ); 1; } or $decode_error = $@;
            
            if ( $decode_error ) {
                return MojoX::JSON::RPC::Dispatcher::Method->new->parse_error(
                    $decode_error );
            }
        }
        $request = {
            method => $params->{method},
            exists $params->{id}    ? ( id     => $params->{id} )   : (),
            defined $decoded_params ? ( params => $decoded_params ) : ()
        };
    }
    else {
        return MojoX::JSON::RPC::Dispatcher::Method->new->invalid_request(
            qq{Invalid method type: $method});
    }

    if ( ref $request ne 'HASH' && ref $request ne 'ARRAY' ) {
        return MojoX::JSON::RPC::Dispatcher::Method->new->invalid_request(
            $request);
    }

    my @methods = ();
METHOD:
    foreach my $obj ( ref $request eq 'HASH' ? $request : @{$request} ) {
        if ( ref $obj ne 'HASH' ) {
            push @methods,
                MojoX::JSON::RPC::Dispatcher::Method->new->invalid_request;
            next METHOD;
        }
        my $method = $obj->{method} if exists $obj->{method};
        if (!(  defined $method
                && $method =~ m/^[A-Za-z_\.][A-Za-z0-9_\.]*$/xms
            )
            )
        {
            push @methods,
                MojoX::JSON::RPC::Dispatcher::Method->new->invalid_request(
                $method);
            next METHOD;
        }
        my $params = $obj->{params} if exists $obj->{params};
        if (   defined $params
            && ref $params ne 'ARRAY'
            && ref $params ne 'HASH' )
        {
            push @methods,
                MojoX::JSON::RPC::Dispatcher::Method->new->invalid_params(
                'NOT array or hash: ' . $params );
            next METHOD;
        }

        ## Hack to support JSON-RPC 1.1
        if (   !( exists $obj->{id} && defined $obj->{id} )
            && exists $obj->{version}
            && defined $obj->{version}
            && $obj->{version} eq '1.1' )
        {
            $obj->{id} = '#';    # TODO: fix this
        }

        ## Create method object
        my $m = MojoX::JSON::RPC::Dispatcher::Method->new(
            method => $method,
            id     => exists $obj->{id} ? $obj->{id} : undef
        );
        if ( !( exists $obj->{id} && defined $obj->{id} ) ) {
            $m->is_notification(1);
        }
        if ( defined $params ) {    # process parameters
            if ( ref $params ne 'ARRAY' && ref $params ne 'HASH' ) {
                $m->invalid_params($params);
            }
            else {
                $m->params($params);
            }
        }
        push @methods, $m;
    }
    return \@methods;
}

sub _handle_methods {
    my ( $self, $methods ) = @_;

    my $log = $self->app->log;

    if ( ref $methods eq 'MojoX::JSON::RPC::Dispatcher::Method' ) {
        if ( $methods->has_error ) {
            return $methods->response;
        }
        $methods = [$methods];
    }
    if ( scalar @{$methods} == 0 ) {    # empty
        return
            MojoX::JSON::RPC::Dispatcher::Method->new->invalid_request
            ->response;
    }

    my @responses;
    my $service = $self->stash('service');
    my $rpcs    = $service->{_rpcs};
METHOD:
    foreach my $m ( ref $methods eq 'ARRAY' ? @{$methods} : $methods ) {
        if ( $m->has_error ) {
            push @responses, $m->response;
            next METHOD;
        }
        my $rpc = $rpcs->{ $m->method }
            if exists $rpcs->{ $m->method };
        my $code_ref = $rpc->{method}
            if defined $rpc && exists $rpc->{method};
        if ( defined $code_ref ) {

            # deal with params and calling
            # pass in svc obj and mojo tx if necessary
            eval {
                $m->result(
                    $code_ref->(
                        exists $rpc->{with_svc_obj}
                            && $rpc->{with_svc_obj} ? $service : (),
                        exists $rpc->{with_mojo_tx}
                            && $rpc->{with_mojo_tx} ? $self->tx : (),
                        exists $rpc->{with_self}
                            && $rpc->{with_self} ? $self : (),
                        defined $m->params
                        ? ref $m->params eq 'ARRAY'
                                ? @{ $m->params }
                                : $m->params
                        : ()
                    )
                );
            };
            if ($@) {
                my $err = $@;
                my $handler = $self->stash('exception_handler');
                if (ref $handler eq 'CODE') {
                    $handler->($self, $err, $m);
                }
                else {
                    $m->internal_error($@);
                }
            }
        }
        else {
            $m->method_not_found( $m->method );
        }
        if ( !$m->is_notification || $m->has_error ) {
            push @responses, $m->response;
        }
    }
    return scalar @responses > 1 ? \@responses : $responses[0];
}

# Translate JSON-RPC error code to HTTP status.
sub _translate_error_code_to_status {
    my ( $self, $code ) = @_;
    my %trans = (
        q{}      => 200,
        '-32600' => 400,
        '-32601' => 404,
        '-32602' => 200,    # wants the user to get the rpc error
        '-32700' => 400
    );
    $code ||= q{};
    return exists $trans{$code} ? $trans{$code} : 500;
}

1;

__END__

=head1 NAME

MojoX::JSON::RPC::Dispatcher - A JSON-RPC 2.0 server for Mojolicious

=head1 SYNOPSIS

    # lib/your-application.pm

    use base 'Mojolicious';
    use MojoX::JSON::RPC::Service;

    sub startup {
        my $self = shift;
        my $svc = MojoX::JSON::RPC::Service->new;

        $svc->register(
            'sum',
            sub {
                my @params = @_;
                my $sum = 0;
                $sum += $_ for @params;
                return $sum;
            }
        );

        $self->plugin(
            'json_rpc_dispatcher',
            services => {
               '/jsonrpc' => $svc
            }
        );
   }

=head1 DESCRIPTION

Using this module you can handle JSON-RPC 2.0 requests within Mojolicious.

=head1 ATTRIBUTES

L<MojoX::JSON::RPC::Dispatcher> inherits all attributes from L<Mojolicious::Controller> and
implements the following attributes.

=head2 C<json>

JSON encoder / decoder

=head2 C<error_code>

Error code.

=head2 C<error_message>

Error message.

=head2 C<error_data>

Error data.

=head2 C<id>

=head1 METHODS

L<MojoX::JSON::RPC::Dispatcher> inherits all methods from L<Mojolicious::Controller> and implements the
following new ones.

=head2 C<call>

Process JSON-RPC call.

=head1 SEE ALSO

L<MojoX::JSON::RPC>, L<Mojolicious::Plugin::JsonRpcDispatcher>

=cut



lib/MojoX/JSON/RPC.pm
package MojoX::JSON::RPC;

use strict;

our $VERSION = '0.09';

1;

__END__

=head1 NAME

MojoX::JSON::RPC - Perl implementation of JSON-RPC 2.0 protocol for Mojolicious

=head1 SYNOPSIS

Server as plugin (L<Mojolicious::Plugin::JsonRpcDispatcher>):

    #!/usr/bin/env perl
    use Mojolicious::Lite;
    use MojoX::JSON::RPC::Service;

    plugin 'json_rpc_dispatcher' => {
        services => {
            '/jsonrpc' => MojoX::JSON::RPC::Service->new->register(
                'sum',
                sub {
                    my @params = @_;
                    my $sum    = 0;
                    $sum += $_ for @params;
                    return $sum;
                }
            )
        }
    };

    app->start;

Client (L<MojoX::JSON::RPC::Client>):

    #!/usr/bin/env perl
    use MojoX::JSON::RPC::Client;

    my $client = MojoX::JSON::RPC::Client->new;
    my $url    = 'http://www.example.com/jsonrpc';
    my $callobj = {
        id      => 1,
        method  => 'sum',
        params  => [ 17, 25 ]
    };

    my $res = $client->call($url, $callobj);

    if($res) {
        if ($res->is_error) { # RPC ERROR
            print 'Error : ', $res->error_message;
        }
        else {
            print $res->result;
        }
    }
    else {
        my $tx_res = $client->tx->res; # Mojo::Message::Response object
        print 'HTTP response '.$tx_res->code.' '.$tx_res->message;
    }

Non-blocking client:

    #!/usr/bin/env perl
    use MojoX::JSON::RPC::Client;

    my $client = MojoX::JSON::RPC::Client->new;
    my $url    = 'http://www.example.com/jsonrpc';
    my $callobj = {
        id      => 1,
        method  => 'sum',
        params  => [ 17, 25 ]
    };

    $client->call($url, $callobj, sub {
        # With callback
        my $res = pop;

        if($res) {
            if ($res->is_error) { # RPC ERROR
                print 'Error : ', $res->error_message;
            }
            else {
                print $res->result;
            }
        }
        else {
           my $tx_res = $client->tx->res; # Mojo::Message::Response object
           print 'HTTP response '.$tx_res->code.' '.$tx_res->message;
        }

        Mojo::IOLoop->stop;
    });

    Mojo::IOLoop->start;

=head1 DESCRIPTION

This module implments a client and a server plugin for JSON-RPC 2.0 for use
with Mojolicious.

This module follows the draft specficiation for JSON-RPC 2.0. More information
can be found at L<http://groups.google.com/group/json-rpc/web/json-rpc-2-0>.

=head1 SEE ALSO

L<Mojolicious::Plugin::JsonRpcDispatcher>, L<MojoX::JSON::RPC::Dispatcher>,
L<MojoX::JSON::RPC::Client>

=head1 AUTHOR

Henry Tang

=head1 CREDITS

=over 2

Igor Afanasyev

Renee Baecker (https://github.com/reneeb)

Vidar Tyldum <vidar@tyldum.com>

=back

=head1 COPYRIGHT & LICENSE

Copyright (C) 2011-2016, Henry Tang.

MojoX::JSON::RPC is provided "as is" and without any express or
implied warranties, including, without limitation, the implied warranties
of merchantibility and fitness for a particular purpose.

This program is free software, you can redistribute it and/or modify it
under the terms of the Artistic License version 2.0.

=cut
